### PR TITLE
Add an option for DreamObjects as external storage

### PR DESCRIPTION
**Pull request by squidboylan**
Basically uses the AmazonS3 code (copied, doesnt use it directly) and fills in some defaults like host.

### CODE REVIEW

**Comment by LukasReschke** [comment]
Thanks a lot for your contribution.

Personally I'm not sure whether we really should add this to core. This means that the expectation is that this is always maintained and support forever in ownCloud.

This is usually quite a big step to do and thus I'd recommend you to implement this as dedicated application and publish it on our appstore. See https://github.com/icewind1991/files_external_ftp for an example on how to do that :)

cc @karlitschek Opinions?

**Comment by karlitschek** [comment]
@LukasReschke Yes. Nowadays external storages like that can be implemented as a separate app. 
@MTRichards What do you think about having this in core?

**Comment by karlitschek** [comment]
confirmed that I got the agreement by the way!

**Comment by smaffulli** [comment]
hi guys, I understand that adding functionalities in core is a big deal. DreamHost will gladly support this with all is necessary to keep it in core, at the same level of Amazon and other storage providers. I'll reach out to you via email with more details.

File: apps/files_external/lib/backend/dreamobjects.php [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+
+set_include_path(get_include_path() . PATH_SEPARATOR .
+	\OC_App::getAppPath('files_external') . '/3rdparty/aws-sdk-php');
+require 'aws-autoloader.php';
+
+use Aws\S3\S3Client;
+use Aws\S3\Exception\S3Exception;
+use Icewind\Streams\IteratorDirectory;
+
+class DreamObjects extends \OC\Files\Storage\Common {
```
Comment (LukasReschke) [inline apps/files_external/lib/backend/dreamobjects.php]: Can you instead extend from `\OC\Files\Storage\AmazonS3`? We should not duplicate the complete Amazon code here. 

Has also the advantage that bugs have only to be fixed at one place etc…

File: apps/files_external/lib/dreamobjects.php [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+
+set_include_path(get_include_path() . PATH_SEPARATOR .
+	\OC_App::getAppPath('files_external') . '/3rdparty/aws-sdk-php');
+require 'aws-autoloader.php';
+
+use Aws\S3\S3Client;
+use Aws\S3\Exception\S3Exception;
+use Icewind\Streams\IteratorDirectory;
+
+class DreamObjects extends \OC\Files\Storage\Common {
```
Comment (LukasReschke) [inline apps/files_external/lib/dreamobjects.php]: This seems duplicated.

Comment (squidboylan) [inline apps/files_external/lib/dreamobjects.php]: yeah must have been, I'll fix that

File: apps/files_external/lib/backend/dreamobjects.php (hunk 1 above)
Comment (squidboylan) [inline apps/files_external/lib/backend/dreamobjects.php]: Yep

**Comment by LukasReschke** [comment]
@squidboylan @smaffulli In addition all changes to core must be completely tested. Can you adjust the core tests to also execute against this backend?

See https://github.com/owncloud/core/tree/master/apps/files_external/tests for the some documentation and the existing code.

**Comment by LukasReschke** [comment]
@squidboylan Mind closing this PR and opening a new one on a branch on this repository? That way our test suite can properly execute against the changes. They are at the moment not executed against forks.

**Comment by squidboylan** [comment]
closing this PR to open a new one as was requested above

**Comment by squidboylan** [comment]
new PR is https://github.com/owncloud/core/pull/23283

**Pull request closed without merging**